companion/lib/Service/Mqtt.ts

```diff
@@ -0,0 +1,201 @@
+		this.port = this.userconfig.getKey('mqtt_port') || 1883
+		this.topic = this.userconfig.getKey('mqtt_topic') || 'companion'
+		this.username = this.userconfig.getKey('mqtt_username') || ''
+		this.password = this.userconfig.getKey('mqtt_password') || ''
+
+		const options: mqtt.IClientOptions = {
+			username: this.username || undefined,
+			password: this.password || undefined,
+			reconnectPeriod: 0, // Disable automatic reconnection from mqtt lib, we handle it with the module restart
+		}
+
+		this.logger.info(`Connecting to MQTT broker at ${this.broker}:${this.port}...`)
+
+		try {
+			this.#client = mqtt.connect(`mqtt://${this.broker}:${this.port}`, options)
+
+			this.#client.on('connect', () => {
+				this.logger.info(`Connected to MQTT broker at ${this.broker}:${this.port}`)
+
+				this.currentState = true
+
+				// Set up event listeners
+				this.startEventListeners()
+
+				// Subscribe to actions topic
+				const actionsTopic = `${this.topic}/commands/#`
+				this.#client!.subscribe(actionsTopic, (err) => {
+					if (err) {
+						this.logger.error(`Failed to subscribe to MQTT topic ${actionsTopic}: ${err.message}`)
+					}
+				})
+
+				// Handle incoming messages
+				this.#client!.on('message', (topic, message) => {
+					this.handleMessage(topic, message)
+					this.logger.debug(`Received MQTT message on topic ${topic}: ${message.toString()}`)
+				})
+			})
+
+			this.#client.on('error', (error) => {
+				this.logger.error(`MQTT error: ${error.message}`)
+				this.stopEventListeners()
+				this.debounceRestart()
+			})
+
+			this.#client.on('close', () => {
+				this.logger.info('Disconnected from MQTT broker')
+				this.stopEventListeners()
+				this.debounceRestart()
+			})
+		} catch (error) {
+			this.logger.error(`Failed to connect to MQTT broker: ${(error as Error).message}`)
+			this.stopEventListeners()
+			this.debounceRestart()
+		}
+	}
+
+	protected close(): void {
+		if (this.#client) {
+			this.#client.end()
+			this.#client = null
+		}
+		this.stopEventListeners()
+	}
+
+	private startEventListeners(): void {
+		// Listen for variable changes
+		this.#serviceApi.on('variables_changed', this.variablesChanged)
+	}
+
+	private stopEventListeners(): void {
+		this.#serviceApi.removeListener('variables_changed', this.variablesChanged)
+	}
+
+	private variablesChanged = (variables: Set<string>): void => {
+		if (this.#client === null) return
+
+		for (const variable of variables) {
+			let [label, name] = variable.split(':')
+
+			if (label == 'internal' && name.startsWith('custom')) {
+				label = 'custom'
+				name = name.substring(7)
+			}
+
+			let value = undefined
+
+			if (label === 'custom') {
+				value = this.#serviceApi.getCustomVariableValue(name)?.toString()
+				if (value === undefined) return
+			} else {
+				value = this.#serviceApi.getConnectionVariableValue(label, name)?.toString()
+				if (value === undefined) return
+			}
+
+			const topic = `${this.topic}/variables/${label}/${name}`
```

Comment:
Just thinking about future proofing, it might be better to do `${this.topic}/variables/${label}/${name}/value`, so that the description or type or something could be added as separate properties later?

Or maybe the message should be a small json blob:
```
{
  "value": "abc"
}
```
to allow for other details to be added with it?